modules/apps/frontend-editor/frontend-editor-ckeditor-web/src/main/resources/META-INF/resources/editor/BalloonEditor.js

```diff
@@ -13,96 +13,176 @@
  */
 
 import PropTypes from 'prop-types';
-import React from 'react';
+import React, {useCallback, useEffect, useRef} from 'react';
 
 import {Editor} from './Editor';
 import DEFAULT_BALLOON_EDITOR_CONFIG from './config/DefaultBalloonEditorConfiguration';
 
 import '../css/main.scss';
 
-const BalloonEditor = ({config = {}, contents, name, ...otherProps}) => {
-	const editorConfig = {
-		...DEFAULT_BALLOON_EDITOR_CONFIG,
-		...config,
-	};
-
-	return (
-		<Editor
-			config={editorConfig}
-			data={contents}
-			name={name}
-			onBeforeLoad={(CKEDITOR) => {
-				CKEDITOR.ADDITIONAL_RESOURCE_PARAMS = {
-					languageId: themeDisplay.getLanguageId(),
-				};
-
-				CKEDITOR.disableAutoInline = true;
-
-				CKEDITOR.getNextZIndex = function () {
-					return CKEDITOR.dialog._.currentZIndex
-						? CKEDITOR.dialog._.currentZIndex + 10
-						: Liferay.zIndex.WINDOW + 10;
-				};
-			}}
-			onInstanceReady={(event) => {
-				const editor = event.editor;
-
-				const editable = editor.editable();
+const EMPTY_OBJECT = {};
+
+const BalloonEditor = React.forwardRef(
```

Comment:
Do we need `forwardRef` here, along with `editorRefsCallback` etc below? I believe we needed it in `ClassicEditor` because `RichText` needed the editor reference. But, there is nothing directly using `BalloonEditor`, right?

Reply:
> I believe we needed it in ClassicEditor because RichText needed the editor reference

It makes sense but I didn't remember it when copying those things, so, I left as is. Does make sense providing a comment inside ClassicEditor a bout this usage?

Reply:
I don't think a comment is needed. `forwardRef` implies parent is controlling the ref. It's like when you have a callback, or an export, you don't need to specify who is consuming it.

Reply:
> I don't think a comment is needed. forwardRef implies parent is controlling the ref

I assume it's okay for React components forwarding their element refs but in our case, the ref contains the editor instance. For me it's not too obvious this use case and I think this description can be enhanced providing the decision behind it but I'm just saying 🤷🏼

Reply:
https://github.com/liferay-frontend/liferay-portal/pull/1663/commits/8c86ae81a123ffc87b8b0ac25631b3dc2093ff1b